flink-connectors/flink-connector-kafka-0.8/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/SimpleConsumerThread.java

```diff
@@ -373,16 +370,28 @@ else if (partitionsRemoved) {
 								keyPayload.get(keyBytes);
 							}
 
-							final T value = deserializer.deserialize(keyBytes, valueBytes, 
-									currentPartition.getTopic(), currentPartition.getPartition(), offset);
-							
-							if (deserializer.isEndOfStream(value)) {
-								// remove partition from subscribed partitions.
-								partitionsIterator.remove();
-								continue partitionsLoop;
-							}
-							
-							owner.emitRecord(value, currentPartition, offset);
+							final Collector<T> collector = new Collector<T>() {
```

Comment:
@haohui hmm this seems a bit odd to me. I think it should be achievable.

```
// the buffer; this can be shared
final List<T> bufferedElements = new LinkedList<>();
// BufferCollector is an implementation of Collector that adds collected elements to bufferedElements; this can be shared
final BufferCollector collector = new BufferCollector<T>(bufferedElements);

...

for (final ConsumerRecord<byte[], byte[]> record : partitionRecords) {
    deserializer.deserialize(
        record.key(), record.value(), record.topic(),
        record.partition(), record.offset(), collector);

    emitRecords(bufferedElements, partitionState, record.offset(), record);

    bufferedElements.clear(); // after the elements for the record have been emitted, empty out the buffer
}
```

Doesn't this work? I haven't really tried this hands-on, so I might be overlooking something. Let me know what you think :)